### DISCUSSION TITLE
Adding a dummy atom to aromatic nitrogen in pyrazole

### DISCUSSION THREAD

**Discussion by xescape** [rdkit/rdkit#5173] [Q&A] [answered]
Hello,

I'm attempting to add a dummy atom to one of the nitrogens in pyrazole, like so:
![add](https://user-images.githubusercontent.com/3373700/162071539-d9859ef4-3f64-4be4-8f54-a8595cf1252c.png)

I'm adding it like this:
```
atom_idx = 2
mol = Chem.MolFromSmiles("C1=CNN=C1")
mol.__sssAtoms = [atom_idx]
display(mol)
```
![image](https://user-images.githubusercontent.com/3373700/162070993-5d40453e-9930-4430-9e2b-1af3ac103b48.png)
```
rw_mol = RWMol(Chem.RemoveAllHs(mol))
dummy_atm = AllChem.Atom(0)
dummy_idx = rw_mol.AddAtom(dummy_atm)
rw_mol.AddBond(atom_idx, dummy_idx, AllChem.BondType.SINGLE)
nitrogen = rw_mol.GetAtomWithIdx(atom_idx)
nitrogen.UpdatePropertyCache()
```

Here I get an error:
`AtomValenceException: Explicit valence for atom # 2 N, 5, is greater than permitted`

and if I look at the product, there's an extra H
'display(rw_mol)'
![image](https://user-images.githubusercontent.com/3373700/162071807-057f5112-6e3d-4368-aa6d-cf9baed8d700.png)

This process seems to work for most situations, so I'm wondering if I'm doing something wrong here. Thank you!

**Comment by greglandrum** [comment, marked as answer]
@xescape the problem here is that aromatic heteroatoms normally have some additional information set to indicate whether or not they have implicit Hs, if you want to add another attachment to them you need to reset this information.
Here's a quick example of how to do that for pyrazole:
```
In [2]: m = Chem.MolFromSmiles('C1=CNN=C1')

In [3]: rwm = Chem.RWMol(m)

In [4]: dummy = Chem.Atom(0)

In [5]: rwm.AddAtom(dummy)
Out[5]: 5

In [6]: rwm.AddBond(2,5,Chem.BondType.SINGLE)
Out[6]: 6

In [7]: rwm.GetAtomWithIdx(2).GetNumExplicitHs()
Out[7]: 1

In [8]: rwm.GetAtomWithIdx(2).GetNoImplicit()
Out[8]: False

In [9]: rwm.GetAtomWithIdx(2).SetNumExplicitHs(0)

In [10]: rwm.GetAtomWithIdx(2).SetNoImplicit(False)

In [11]: rwm.UpdatePropertyCache()

In [12]: Chem.MolToSmiles(rwm)
Out[12]: '*n1cccn1'
```
The two important calls are the ones in lines 9 and 10.

**Reply by xescape** [reply]
Yes, setting the explicit and implicit Hs solved my problem. Thanks for your help @greglandrum !
